design/restore-with-EnableAPIGroupVersions-feature.md

```diff
@@ -0,0 +1,211 @@
+
+## Detailed Design
+
+There are six objectives to achieve the above stated goals:
+
+1. Determine if the APIGroupVersionsFeatureFlag feature flag is enabled and Backup Objects use Status.FormatVersion 1.1.0.
+1. List the backed up API group versions.
+1. List the API group versions supported by the target cluster.
+1. Get the user-defined version priorities.
+1. Use a priority system to determine which version to restore. The source preferred version will be the default if the priorities fail.
+1. Modify the paths to the backup files in the tarball in the resource restore process.
+
+### Objective 1: Determine if the APIGroupVersionsFeatureFlag feature flag is enabled and Backup Objects use Status.FormatVersion 1.1.0
+
+For restore to be able to choose from multiple supported backed up versions, the feature flag must have been enabled during the restore processes. Backup objects must also have [Status.FormatVersion == "1.1.0"](https://github.com/vmware-tanzu/velero/blob/a1e182e723a8c5f6d4175d8db2361233a94d2502/pkg/backup/backup.go#L58).
+
+The reason for checking for the feature flag during restore is to ensure the user would like to restore a version that might not be the source cluster preferred version.
+
+The reason for checking `Status.FormatVersion` is to ensure the changes made by this proposed design is backward compatible. Only with Velero version 1.4 and forward was Format Version 1.1.0 used to structure the backup directories. Format Version 1.1.0 is required for the restore process proposed in this design doc to work. Before v1.4, the backed up files were in a directory structure that will not be recognized by the proposed code changes. In this case, restore should not attempt to restore from multiple versions as they will not exist.
+
+Checking if the feature flag is enabled during restore is straightforward using `features.IsEnabled(velerov1api.APIGroupVersionsFeatureFlag)`.
+
+Checking the `Status.FormatVersion` can be done indirectly by checking the directory structure and version directory names. For backups generated using Format Version 1.1.0, there will always be a version directory with `-preferredversion` in the directory name. Only one resource will need to be checked to determine if the feature flag was enabled during backup.
+
+To help with the format version check, a new parser object can be created using the construction function `NewParser(...)` and we can take advantage of the `archive` package's `Parser` type. A method for the parser object will be created that has the signature `UsedFormatVersion(dir string) (bool, error)`. The method will loop through the version directories inside one of the resource.group directories and determine if a directory with `-preferredversion` was found.
```

Comment:
Similar question - is this paragraph, and the `Parser` needed since you're using the `Status.FormatVersion` filed directly instead of inferring it based on directory structure?

Reply:
I'd be able to get rid of `UsedFormatVersion` function, but I'll still need `Parser` to read the version directories, e.g. `v1-preferredversion`, `v2`, etc.